R/RLearner_regr_gstat.R

```diff
@@ -0,0 +1,164 @@
+      alpha = 0, beta.variogram = 0, cressie = FALSE, dX = 0, boundaries = 0, cloud = FALSE,
+      trend.beta = NULL, debug.level = 1, cross = TRUE, map = FALSE, projected = TRUE, lambda = 1.0,
+      verbose = FALSE, covariogram = FALSE, PR = FALSE, pseudo = -1,
+      fit.sills = TRUE, fit.ranges = TRUE, fit.method = 7, warn.if.neg = FALSE, fit.kappa = FALSE,
+      psill = NA, range = NA, kappa = 0.5, Err = 0),
+    properties = c("numerics", "factors" , "se", "weights", "missings"),
+    name = "Multivariable Geostatistical Prediction And Simulation",
+    short.name = "gstat",
+    note = "The default prediction (when passing no arguments) is IDW (inverse distance weighted).
+    The learner handles gstat variogram autofitting functionnality presented in this post https://www.r-spatial.org/r/2016/02/14/gstat-variogram-fitting.html.\n
+    Manual fitting example : lrn.man = makeLearner(cl = 'regr.gstat', id = 'manual', psill = 1, model = 'Sph', range = 900, nugget = 2, locations = ~x+y).\n
+    Automatic fitting example : lrn.auto = makeLearner(cl = 'regr.gstat', id = 'auto', psill = c('Sph','Exp','Gau', 'Mat'), locations = ~x+y)"
+  )
+}
+
+#' @export
+# https://stackoverflow.com/questions/19075331/passing-a-function-argument-to-other-arguments-which-are-functions-themselves
+# https://stackoverflow.com/questions/16774946/passing-along-ellipsis-arguments-to-two-different-functions
+trainLearner.regr.gstat = function(.learner, .task, .subset, .weights = NULL, ...) {
+  dots = list(...)
+  # https://stackoverflow.com/questions/11885207/get-all-parameters-as-list
+  variogram.names = names(formals(gstat::variogram)) # FIXME cannot retrieve the S3 method for formula arguments
+  variogram.names = c("object", "locations", "beta.variogram")
+  #variogram.names = replace(names(formals(gstat::variogram)), variogram.names == "beta", "beta.variogram")
+  fit.variogram.names = names(formals(gstat::fit.variogram))
+  gstat.names = replace(names(formals(gstat::gstat)), names(formals(gstat::gstat)) == "beta", "beta.gstat")
+
+  d = getTaskData(.task, .subset)
+  f = getTaskFormula(.task, explicit.features = TRUE)
+  # remove location vars as they are handled by gstat - https://stackoverflow.com/questions/40308944/removing-offset-terms-from-a-formula
+  f = update(f, .~.-y-x) # FIXME should be the params entered in locations arg
+  # check if a variogram model is passed
+  if (!is.na(dots$psill)) {
+    # build the samples variogram
+    v = do.call(gstat::variogram, c(list(object = f, data = d), dots[ names(dots) %in% variogram.names] ))
+    # fit the variogram model
+    fit = do.call(gstat::fit.variogram,
+      c(list(object = v,
+        model = gstat::vgm(psill = dots$psill,
+          model = dots$model,
+          range = dots$range,
+          nugget = dots$nugget)),
+        dots[names(dots) %in% fit.variogram.names[fit.variogram.names != "model"]])
```

Comment:
Why is `model` filtered out here?

Reply:
See comment above. `gstat::vgm` and `gstat::fit.variogram` both require a `model` argument. The `model` argument you pass to the learner is the one that goes to `gstat::vgm` which output is the `model` argument required for `gstat::fit.variogram`. So while constructing `fit.variogram`, I must remove the model argument stored in `...` (`dots`) because it only as a meaning in the `gstat::vgm` function.